2-Authorization-I/1-call-graph/SPA/src/graph.js

```diff
@@ -0,0 +1,116 @@
+import { Client } from '@microsoft/microsoft-graph-client';
+
+import { msalInstance } from "./index";
+
+/**
+ * This class implements the IAuthenticationProvider interface, which allows a custom auth provider to be
+ * used with the Graph client. See: https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/dev/src/IAuthenticationProvider.ts
+ */
+class MyAuthenticationProvider {
+
+    userAccount;
+    requiredScopes;
+    interactionType;
+
+    /**
+     * @param {object} userAccount: user account object to be used when attempting silent token acquisition  
+     * @param {array} requiredScopes: array of scopes required for this resource endpoint
+     * @param {string} interactionType: type of interaction to fallback to when silent token acquisition fails 
+     */
+    constructor(userAccount, requiredScopes, interactionType) {
+        this.userAccount = userAccount;
+        this.requiredScopes = requiredScopes;
+        this.interactionType = interactionType;
+    }
+
+    /**
+     * This method will get called before every request to the ms graph server
+     * This should return a Promise that resolves to an accessToken (in case of success) or rejects with error (in case of failure)
+     * Basically this method will contain the implementation for getting and refreshing accessTokens
+     */
+    getAccessToken() {
+        return new Promise(async (resolve, reject) => {
+            let response;
+
+            try {
+                response = await msalInstance.acquireTokenSilent({
+                    account: this.userAccount,
+                    scopes: this.requiredScopes
+                });
+
+                if (response.accessToken) {
+                    resolve(response.accessToken);
+                } else {
+                    reject(Error('Failed to acquire an access token'));
+                }
+            } catch (error) {
+                // in case if silent token acquisition fails, fallback to an interactive method
+                if (error instanceof InteractionRequiredAuthError) {
+                    switch (this.interactionType) {
+                        case InteractionType.Popup:
+
+                            response = await msalInstance.acquireTokenPopup({
+                                scopes: this.requiredScopes
+                            });
+
+                            if (response.accessToken) {
+                                resolve(response.accessToken);
+                            } else {
+                                reject(Error('Failed to acquire an access token'));
+                            }
+                            break;
+
+                        case InteractionType.Redirect:
+                            msalInstance.acquireTokenRedirect({
+                                scopes: this.requiredScopes
+                            });
+                            
+                            msalInstance.addEventCallback((message) => {
```

Comment:
This won't ever be hit as `acquireTokenRedirect` will leave the page. You may not need this at all as `acquireTokenSilent` should succeed after returning from the redirect. If you do need the direct redirect response you should move this as far up as you can to ensure the event callback gets registered before the response is processed.

Reply:
Ah, that explains why it was working without registering a callback. I moved it up and was able to register, but it doesn't get called as `acquireTokenSilent` succeeds before when returning from redirect. Removing then 👍